[JIT] Regularize tensorexpr fuser strategy with other fusers

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44995 Clean up sortReverseTopological copy pasta
* **#44972 [JIT] Regularize tensorexpr fuser strategy with other fusers**
* #44238 Refactor subgraph merging

Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 

The basic strategy is: 
- iterate until you find a fusible node
- try to merge the nodes inputs, whenever a succesful merge occurs restart at the beginning of the nodes inputs
- after you've exhausted a node, continue searching the block for fusion opportunities from the node 
- continue doing this on the block until we go through an iteration without an succesful merges

Since we create the fusion groups once, and only re-specialize within the fusion groups, we should be running this very infrequently (only re-triggers when we fail undefinedness specializations). Also bc it's the same algorithm as the existing fuser it is unlikely to cause a regression.

Differential Revision: [D23821581](https://our.internmc.facebook.com/intern/diff/D23821581)

Commit:
Regularize tensorexpr fuser strategy with other fusers

[ghstack-poisoned]

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"




[ghstack-poisoned]

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"


Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 





[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -378,43 +379,44 @@ class TensorExprFuser {
+      changed = true;
     }
 
-    updateQueue(fusion_group, queue, visited_nodes);
```
You can probably garbage collect the updateQueue function since it appears no longer used.

bertmaher commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -424,68 +426,82 @@ class TensorExprFuser {
       // If merging suceeded, save the merged group as the "previous" fusion
       // group so that we can try to merge the next one into it.
-      if (prev_fusion_group) {
+
```
Do you know if this logic to merge adjacent fusion groups is still necessary?  It seems like the iterative fuser approach with scanNode would pick these cases up, but maybe I'm not being creative enough :-D

eellison replied:
Good point

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"


Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 





[ghstack-poisoned]

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"


Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 

The basic strategy is: 
- iterate until you find a fusible node
- try to merge the nodes inputs, whenever a succesful merge occurs restart at the beginning of the nodes inputs
- after you've exhausted a node, continue searching the block for fusion opportunities from the node 
- continue doing this on the block until we go through an iteration without an succesful merges

Since we create the fusion groups once, and only re-specialize within the fusion groups, we should be running this very infrequently (only re-triggers when we fail undefinedness specializations). Also bc it's the same algorithm as the existing fuser it is unlikely to cause a regression. 





[ghstack-poisoned]

bertmaher reviewed (approved):
Looks great!  And I love the extremely clear unit test.

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"


Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 

The basic strategy is: 
- iterate until you find a fusible node
- try to merge the nodes inputs, whenever a succesful merge occurs restart at the beginning of the nodes inputs
- after you've exhausted a node, continue searching the block for fusion opportunities from the node 
- continue doing this on the block until we go through an iteration without an succesful merges

Since we create the fusion groups once, and only re-specialize within the fusion groups, we should be running this very infrequently (only re-triggers when we fail undefinedness specializations). Also bc it's the same algorithm as the existing fuser it is unlikely to cause a regression. 





[ghstack-poisoned]

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"


Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 

The basic strategy is: 
- iterate until you find a fusible node
- try to merge the nodes inputs, whenever a succesful merge occurs restart at the beginning of the nodes inputs
- after you've exhausted a node, continue searching the block for fusion opportunities from the node 
- continue doing this on the block until we go through an iteration without an succesful merges

Since we create the fusion groups once, and only re-specialize within the fusion groups, we should be running this very infrequently (only re-triggers when we fail undefinedness specializations). Also bc it's the same algorithm as the existing fuser it is unlikely to cause a regression.

Differential Revision: [D23821581](https://our.internmc.facebook.com/intern/diff/D23821581)

[ghstack-poisoned]

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"


Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 

The basic strategy is: 
- iterate until you find a fusible node
- try to merge the nodes inputs, whenever a succesful merge occurs restart at the beginning of the nodes inputs
- after you've exhausted a node, continue searching the block for fusion opportunities from the node 
- continue doing this on the block until we go through an iteration without an succesful merges

Since we create the fusion groups once, and only re-specialize within the fusion groups, we should be running this very infrequently (only re-triggers when we fail undefinedness specializations). Also bc it's the same algorithm as the existing fuser it is unlikely to cause a regression.

Differential Revision: [D23821581](https://our.internmc.facebook.com/intern/diff/D23821581)

[ghstack-poisoned]

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"


Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 

The basic strategy is: 
- iterate until you find a fusible node
- try to merge the nodes inputs, whenever a succesful merge occurs restart at the beginning of the nodes inputs
- after you've exhausted a node, continue searching the block for fusion opportunities from the node 
- continue doing this on the block until we go through an iteration without an succesful merges

Since we create the fusion groups once, and only re-specialize within the fusion groups, we should be running this very infrequently (only re-triggers when we fail undefinedness specializations). Also bc it's the same algorithm as the existing fuser it is unlikely to cause a regression.

Differential Revision: [D23821581](https://our.internmc.facebook.com/intern/diff/D23821581)

[ghstack-poisoned]

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"


Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 

The basic strategy is: 
- iterate until you find a fusible node
- try to merge the nodes inputs, whenever a succesful merge occurs restart at the beginning of the nodes inputs
- after you've exhausted a node, continue searching the block for fusion opportunities from the node 
- continue doing this on the block until we go through an iteration without an succesful merges

Since we create the fusion groups once, and only re-specialize within the fusion groups, we should be running this very infrequently (only re-triggers when we fail undefinedness specializations). Also bc it's the same algorithm as the existing fuser it is unlikely to cause a regression.

Differential Revision: [D23821581](https://our.internmc.facebook.com/intern/diff/D23821581)

[ghstack-poisoned]

Commit:
Update on "[JIT] Regularize tensorexpr fuser strategy with other fusers"


Previously, our fusion strategy would be: 
- start at the end of the block, find a fusable node
- iteratively try to merge inputs into the fusion group, sorted topologically 

This strategy works pretty well, but has the possibility of missing fusion groups. See my attached test case for an example where we wouldn't find all possible fusion groups. @bertmaher found an example of a missed fusion groups in one of our rnn examples (jit_premul) that caused a regression from the legacy fuser. 

Here, I'm updating our fusion strategy to be the same as our other fusion passes - create_autodiff_subgraphs, and graph_fuser.cpp. 

The basic strategy is: 
- iterate until you find a fusible node
- try to merge the nodes inputs, whenever a succesful merge occurs restart at the beginning of the nodes inputs
- after you've exhausted a node, continue searching the block for fusion opportunities from the node 
- continue doing this on the block until we go through an iteration without an succesful merges

Since we create the fusion groups once, and only re-specialize within the fusion groups, we should be running this very infrequently (only re-triggers when we fail undefinedness specializations). Also bc it's the same algorithm as the existing fuser it is unlikely to cause a regression.

Differential Revision: [D23821581](https://our.internmc.facebook.com/intern/diff/D23821581)

[ghstack-poisoned]
